### ISSUE TITLE
allow `false` as child nodes

### ISSUE THREAD

**Issue by donpark** [Matt-Esch/virtual-dom#326]
Common expression I use to inline optional child is:

    h('.foo', [
        someFlag && someTest() && h('span', 'optional element'),
        h('div', ...)
    ]);

If both `someFlag` and `someTest()` function returns true, optional element gets inserted.
If not, `false` results which `virtual-dom` currently throws Exception on.
Alternative is to build children array separately and test before inserting into position explicitly.

It would be super cool if virtual-dom just ignored values it can't make sense out of.

**Comment by OliverJAsh** [comment]
I would also like to see this.

**Comment by Raynos** [comment]
Expr ? Node : null works today

You do not really want to filter out false values because it moves Dom nodes around unnecessarily

**Comment by OliverJAsh** [comment]
> it moves Dom nodes around unnecessarily

Can you elaborate?

**Comment by donpark** [comment]
> Expr ? Node : null works today

I know but I think `foo && bar && h(...)` is clearer than `foo ? bar ? h(...) : null : null`.

> it moves Dom nodes around unnecessarily

I also need in help in understanding this. If you mean `h(...)` will get called unnecessarily, I don't think it will unless both `foo` and `bar` expressions return true.

**Comment by Raynos** [comment]
If you do a something like

```js
h('div', [
  h('div', 'one'),
  expr ? h('div', 'two') : null,
  reallyExpensiveThing()
])
```

Then the diff when `expr` changes is simply an insertDOMNode() or removeDOMNode()

If you do something like

```js
h('div', [
  h('div', 'one'),
  expr && h('div', 'two'),
  reallyExpensiveThing()
].filter(Boolean))
```

Then when `expr` changes the amount of children changes between 2 & 3.

Whenever there is a mismatch in the length of children we remove node 2 and rebuild it and re-insert it at node 1.

The `null` has a very explicit semantics of "insert a null dom node".

virtual-dom NEVER moves dom nodes without the key feature, only if you use the key feature will virtual-dom move dom nodes around.

The only operations it supports are insert() and remove() at indexes, which is why that null placeholder is important. 

Basically whenever possible you want `children.length` to be a stable number to avoid churning the DOM.

**Comment by donpark** [comment]
Why would anyone apply `.filter(Boolean)` there? Obviously what I don't know exceeds what I know in virtual-dom use-cases.

Anyway, without `.filter(Boolean)`:

    var h = require('virtual-dom/h');
    var diff = require('virtual-dom/diff');
    var expr = true;
    function reallyExpensiveThing() { return h('span'); }

    var r1 = h('div', [ h('div', 'one'), true && h('div', 'two'), reallyExpensiveThing() ]);
    var r2 = h('div', [ h('div', 'one'), false && h('div', 'two'), reallyExpensiveThing() ]);
    var d12 = diff(r1, r2);

    var r3 = h('div', [ h('div', 'one'), true ? h('div', 'two') : null, reallyExpensiveThing() ]);
    var r4 = h('div', [ h('div', 'one'), false ? h('div', 'two') : null, reallyExpensiveThing() ]);
    var d34 = diff(r3, r4);

    console.log("diff(d12, d34)", diff(d12, d34));

returns:

    diff(d12, d34) { a: 
       { '3': VirtualPatch { type: 2, vNode: [Object], patch: [Object] },
         '5': VirtualPatch { type: 7, vNode: [Object], patch: undefined },
         a: 
          VirtualNode {
            tagName: 'DIV',
            properties: {},
            children: [Object],
            key: undefined,
            namespace: null,
            count: 5,
            hasWidgets: false,
            hasThunks: false,
            hooks: undefined,
            descendantHooks: false } } }

From `diff` perspective, difference appears to be just between `[Object]` vs `undefined`.

Anyway, I just created this use-cases because what annoyed me may annoy others.
I'm sure you'll know what's best for virtual-dom, a project I'm glad to have in any shape or form.

**Comment by Raynos** [comment]
I'll let @matt-esch chime in.
